File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/background/providers/impl/GoogleLocationProvider.java
package se.gustavkarlsson.aurora_notifier.android.background.providers.impl;

import android.location.Location;
import android.support.annotation.NonNull;
import android.util.Log;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationServices;

import se.gustavkarlsson.aurora_notifier.android.R;
import se.gustavkarlsson.aurora_notifier.android.background.providers.LocationProvider;
import se.gustavkarlsson.aurora_notifier.android.util.UserFriendlyException;

import static java.util.concurrent.TimeUnit.MILLISECONDS;


public class GoogleLocationProvider implements LocationProvider {
	private static final String TAG = GoogleLocationProvider.class.getSimpleName();

	private final GoogleApiClient googleApiClient;

	public GoogleLocationProvider(GoogleApiClient googleApiClient) {
		this.googleApiClient = googleApiClient;
	}

	@Override
	public Location getLocation(long timeoutMillis) {
		Log.i(TAG, "Connecting to Google Play Services...");
		try {
			ConnectionResult connectionResult = googleApiClient.blockingConnect(timeoutMillis, MILLISECONDS);
			if (!connectionResult.isSuccess()) {
				if (connectionResult.getErrorCode() == ConnectionResult.TIMEOUT) {
					throw new UserFriendlyException(R.string.update_took_too_long, "Connecting to Google API timed out after " + timeoutMillis + "ms");
				}
				throw new UserFriendlyException(R.string.could_not_connect_to_google_play_services, createErrorMessage(connectionResult));
			}
			Log.d(TAG, "Successfully connected to Google Play Services");
			Log.d(TAG, "Getting location...");
			Location location = LocationServices.FusedLocationApi.getLastLocation(googleApiClient);
			Log.d(TAG, "Location is: " + location);
			if (location == null) {
				throw new UserFriendlyException(R.string.could_not_determine_location, "Location API returned null");
			}
			return location;
		} catch (SecurityException e) {
			Log.w(TAG, "Location permission missing", e);
			throw new UserFriendlyException(R.string.location_permission_missing, e);
		} finally {
			if (googleApiClient.isConnected()) {
				Log.d(TAG, "Disconnecting from Google Play Services");
				googleApiClient.disconnect();
			}
		}
	}

	@NonNull
	private static String createErrorMessage(ConnectionResult connectionResult) {
		return "Could not connect to Google Play Services" +
                ". Error code: " + connectionResult.getErrorCode() +
                ". Error message: " + connectionResult.getErrorMessage();
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/background/providers/impl/GoogleLocationProvider.kt
package se.gustavkarlsson.skylight.android.background.providers.impl

import android.location.Location
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationServices
import dagger.Reusable
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.debug
import org.jetbrains.anko.info
import org.jetbrains.anko.warn
import org.threeten.bp.Duration
import se.gustavkarlsson.skylight.android.R
import se.gustavkarlsson.skylight.android.background.providers.LocationProvider
import se.gustavkarlsson.skylight.android.util.UserFriendlyException
import java.util.concurrent.TimeUnit.MILLISECONDS
import javax.inject.Inject

@Reusable
class GoogleLocationProvider
@Inject
constructor(
		private val googleApiClient: GoogleApiClient
) : LocationProvider, AnkoLogger {

    override fun getLocation(timeout: Duration): Location {
        info("Connecting to Google Play Services...")
        try {
            val connectionResult = googleApiClient.blockingConnect(timeout.toMillis(), MILLISECONDS)
			handleFailure(connectionResult, timeout)
            debug("Successfully connected to Google Play Services")
            debug("Getting location...")
            val location = LocationServices.FusedLocationApi.getLastLocation(googleApiClient)
            debug("Location is: $location")
            if (location == null) {
                throw UserFriendlyException(R.string.error_could_not_determine_location, "Location API returned null")
            }
            return location
        } catch (e: SecurityException) {
            warn("Location permission missing", e)
            throw UserFriendlyException(R.string.error_location_permission_missing, e)
        } finally {
            if (googleApiClient.isConnected) {
                debug("Disconnecting from Google Play Services")
                googleApiClient.disconnect()
            }
        }
    }

	private fun handleFailure(connectionResult: ConnectionResult, timeout: Duration) {
		if (!connectionResult.isSuccess) {
			if (connectionResult.errorCode == ConnectionResult.TIMEOUT) {
				throw UserFriendlyException(R.string.error_updating_took_too_long, "Connecting to Google API timed out after ${timeout.toMillis()}ms")
			}
			throw UserFriendlyException(R.string.error_could_not_connect_to_google_play_services, createErrorMessage(connectionResult))
		}
	}

	private fun createErrorMessage(connectionResult: ConnectionResult): String {
		return "Could not connect to Google Play Services. Error code: ${connectionResult.errorCode}. Error message: ${connectionResult.errorMessage}"
	}
}
